samples/fpgaperf_counter/fpgaperf_counter.c

```diff
@@ -0,0 +1,639 @@
+		OPAE_MSG("Failed to unlock perf mutex");
+		return FPGA_EXCEPTION;
+	}
+	return FPGA_OK;
+out:
+	opae_mutex_unlock(res, &fpga_perf_lock);
+	return FPGA_EXCEPTION;
+}
+
+
+fpga_result fpga_Perf_Events(char* perf_sysfs_path)
+{
+	fpga_result ret 	= FPGA_OK;
+	struct udev *udev 	= NULL;
+	struct udev_device *dev = NULL;
+	int fd 			= 0;
+	int grpfd		= 0;
+	int res 		= 0;
+	uint64_t loop 		= 0;
+	struct perf_event_attr pea;
+	
+	if (!perf_sysfs_path) {
+		OPAE_ERR("Invalid input parameters");
+		return FPGA_INVALID_PARAM;
+	}
+	/* create udev object */
+	udev = udev_new();
+	if (!udev) {
+		OPAE_ERR("Cannot create udev context");
+		return FPGA_EXCEPTION;
+	}
+	dev = udev_device_new_from_syspath(udev, perf_sysfs_path);
+	if (!dev) {
+		OPAE_ERR("Failed to get device");
+		udev_unref(udev);
+		return FPGA_EXCEPTION;
+	}	
+	if (opae_mutex_lock(res, &fpga_perf_lock)) {
+		OPAE_MSG("Failed to lock perf mutex");
+		goto err;
+	}
+	const char * ptr = udev_device_get_sysattr_value(dev, "cpumask");
+	if (ptr)
+		PARSE_MATCH_INT(ptr, 0, g_fpga_perf->cpumask, 10);
+
+	ptr = udev_device_get_sysattr_value(dev, "type");
+	if (ptr)
+		PARSE_MATCH_INT(ptr, 0, g_fpga_perf->type, 10);
+
+	if (opae_mutex_unlock(res, &fpga_perf_lock)) {
+		OPAE_MSG("Failed to unlock perf mutex");
+		goto err;
+	}
+	/* parse the format value */
+	ret = parse_Perf_Format(dev);
+	if (ret != FPGA_OK)
+		goto err;
+	/* parse the event value */
+	ret = parse_Perf_Event(dev);
+	if (ret != FPGA_OK)
+		goto err;
+
+	/* initialize the pea structure to 0 */
+	memset(&pea, 0, sizeof(struct perf_event_attr));
+	
+	if (opae_mutex_lock(res, &fpga_perf_lock)) {
+		OPAE_MSG("Failed to lock perf mutex");
+		goto err;
+	}
+
+	for (loop = 0; loop < g_fpga_perf->num_perf_events; loop++) {
+
+		if (g_fpga_perf->perf_events[0].fd <= 0)
+			grpfd = -1;
+		else
+			grpfd = g_fpga_perf->perf_events[0].fd;
+		
+		if (!g_fpga_perf->perf_events[loop].config)
+			continue;
+
+		pea.type = g_fpga_perf->type;
+		pea.size = sizeof(struct perf_event_attr);
+		pea.config = g_fpga_perf->perf_events[loop].config;
+		pea.disabled = 1;
+		pea.inherit = 1;
+		pea.sample_type = PERF_SAMPLE_IDENTIFIER;
+		pea.read_format = PERF_FORMAT_GROUP | PERF_FORMAT_ID;
+		fd = syscall(__NR_perf_event_open, &pea, -1, g_fpga_perf->cpumask, grpfd, 0);
+		if (fd == -1) {
+			OPAE_ERR("Error opening leader %llx\n", pea.config);
+			ret = FPGA_EXCEPTION;
+			goto out;
+		} else {
+			g_fpga_perf->perf_events[loop].fd = fd;
+			if (ioctl(g_fpga_perf->perf_events[loop].fd, PERF_EVENT_IOC_ID,
+						&g_fpga_perf->perf_events[loop].id) == -1) {
+				OPAE_ERR("PERF_EVENT_IOC_ID ioctl failed: %s",
+						strerror(errno));
+				ret = FPGA_EXCEPTION;
+				goto out;
+			}
+		}
+	}
+	if (ioctl(g_fpga_perf->perf_events[0].fd, PERF_EVENT_IOC_RESET,
+				PERF_IOC_FLAG_GROUP) == -1) {
+		OPAE_ERR("PERF_EVENT_IOC_RESET ioctl failed: %s", strerror(errno));
+		ret = FPGA_EXCEPTION;
+		goto out;
+	}
+out:
+	udev_unref(udev);
+	udev_device_unref(dev);
```

Comment:
I'm not sure it it matters or not but I would release resources in the opposite order that they were acquired, especially if one resource depends on another acquired before it.

Reply:
changed the order of releasing resources